make sure all IDE analyzers implement IBuildInAnalyzer

heejaechang opened the pull request:
IDEAnalyzer does this special thing that compiler analyzer can't do.
that is reporting all its diagnostic descriptor as hidden and change its
severity on runtime.

engine has special treatment for this behavior only if the analyzer implements
IBuiltInAnalyzer. otherwise, the analyzer will be treated as normal
compiler analyzer and things won't work as expected

...

**Customer scenario**

Customer opens files to edit.  Customer seems 'code fix' messages for this file show up in the Error List in the Messages column.  Customer then closes the file and goes onto other files to do work.   However, the messages in the error-list never go away.  The list gets longer and longer the more work the customer does and the more files they go and edit.  This leads to a very noisy and annoying experience.   It's also feels very broken as normal ErrorList errors and Warnings do not behave this way, but nicely go away as you close files.

**Bugs this fixes:** 

https://github.com/dotnet/roslyn/issues/16415

**Workarounds, if any**

there is no workaround

**Risk**

low since this was existing behavior before recent refactoring.

**Performance impact**

no impact since this was existing behavior before recent change.

**Is this a regression from a previous update?**

Yes.

**Root cause analysis:**

some IDE analyzer forgot to implement IBuiltInAnalyzer interface.

**How was the bug found?**

dogfooding.

heejaechang commented:
@srivatsn @mavasani @panopticoncentral @dpoeschl @CyrusNajmabadi @Pilchie can you take a look?

heejaechang commented:
we probably want to take it to RC3?

CyrusNajmabadi commented on src/Features/Core/Portable/UseAutoProperty/AbstractUseAutoPropertyAnalyzer.cs:
```diff
@@ -29,6 +30,9 @@ protected AbstractUseAutoPropertyAnalyzer()
         }
 
+        public override bool OpenFileOnly(Workspace workspace) => false;
+        public override DiagnosticAnalyzerCategory GetAnalyzerCategory() => DiagnosticAnalyzerCategory.ProjectAnalysis;
```
what does 'ProjectAnalysis' mean?

heejaechang replied:
it means it uses CompilationEnd analysis

CyrusNajmabadi commented on src/Features/Core/Portable/UseCoalesceExpression/AbstractUseCoalesceExpressionDiagnosticAnalyzer.cs:
```diff
@@ -24,6 +25,9 @@ protected AbstractUseCoalesceExpressionDiagnosticAnalyzer()
         }
 
+        public override bool OpenFileOnly(Workspace workspace) => false;
+        public override DiagnosticAnalyzerCategory GetAnalyzerCategory() => DiagnosticAnalyzerCategory.SemanticSpanAnalysis;
```
IIRC, we could not use SemanticSpan here.  That's because the analyzer engine will not update 'AdditionalLocations' on Diagnostics properly if you use that.

heejaechang replied:
sure

CyrusNajmabadi commented on src/Features/Core/Portable/UseCoalesceExpression/AbstractUseCoalesceExpressionForNullableDiagnosticAnalyzer.cs:
```diff
@@ -22,12 +22,15 @@ internal abstract class AbstractUseCoalesceExpressionForNullableDiagnosticAnalyz
         }
 
+        public override bool OpenFileOnly(Workspace workspace) => false;
+        public override DiagnosticAnalyzerCategory GetAnalyzerCategory() => DiagnosticAnalyzerCategory.SemanticSpanAnalysis;
```
same here.

heejaechang replied:
sure

CyrusNajmabadi commented on src/Features/Core/Portable/UseCollectionInitializer/AbstractUseCollectionInitializerDiagnosticAnalyzer.cs:
```diff
@@ -28,7 +29,8 @@ internal abstract partial class AbstractUseCollectionInitializerDiagnosticAnalyz
     {
-        public bool OpenFileOnly(Workspace workspace) => false;
+        public override bool OpenFileOnly(Workspace workspace) => false;
+        public override DiagnosticAnalyzerCategory GetAnalyzerCategory() => DiagnosticAnalyzerCategory.SemanticSpanAnalysis;
```
same here.

heejaechang replied:
sure

CyrusNajmabadi commented on src/Features/Core/Portable/UseNullPropagation/AbstractUseNullPropagationDiagnosticAnalyzer.cs:
```diff
@@ -33,6 +33,9 @@ protected AbstractUseNullPropagationDiagnosticAnalyzer()
         }
 
+        public override bool OpenFileOnly(Workspace workspace) => false;
+        public override DiagnosticAnalyzerCategory GetAnalyzerCategory() => DiagnosticAnalyzerCategory.SemanticSpanAnalysis;
```
same here.

heejaechang replied:
sure

CyrusNajmabadi commented on src/Features/Core/Portable/UseNullPropagation/AbstractUseNullPropagationDiagnosticAnalyzer.cs:
```diff
@@ -33,6 +33,9 @@ protected AbstractUseNullPropagationDiagnosticAnalyzer()
         }
 
+        public override bool OpenFileOnly(Workspace workspace) => false;
+        public override DiagnosticAnalyzerCategory GetAnalyzerCategory() => DiagnosticAnalyzerCategory.SemanticDocumentAnalysis;
```
what does SemanticDocumentAnalysis mean.

heejaechang replied:
http://source.roslyn.io/#Microsoft.CodeAnalysis.Features/Diagnostics/DiagnosticAnalyzerCategory.cs,b9f70de523d71032

heejaechang commented:
@Pilchie I think we want to take this to RC3.

heejaechang commented:
I did the work to help IDE team. if IDE want to do it differently, you guys can either modify this PR or send out new one. but regardless, I think we want to fix this for RC3.

Pilchie commented:
@heejaechang to make that call, I'll need a more detailed description of the impact to customers of this issue.  Your description above is much too vague.

heejaechang commented:
basically, diagnostics might run in OOP when analyzer is not supposed to run in OOP (since it is open file only). so options might not synched properly making option not working.

open file only diagnostics might not get removed when the file closed. so diagnostics left out in error list.

with FSA on, some analyzer might never run on closed file since it always return hidden descriptor regardless of options.

CyrusNajmabadi commented:
👍

Pilchie commented:
I'm still scared to take this to RC.3 tomorrow, given it's the very last day.

@heejaechang You mention this is a regression - when was it introduced?  Does RC.2 have the same set of problems?  If so, I'd rather keep the existing behavior until we have time to test these changes more deeply.

heejaechang commented:
@Pilchie okay, if you feels that's more comfortable.

heejaechang commented:
@Pilchie but at least I can check this in to master?

Pilchie commented:
@heejaechang You didn't answer my question - what is the state of RC.2 with respect to this change?

Regarding master - it is also in escrow.  Once this change has been appropriately tested, we can take it to shiproom.

heejaechang commented:
@Pilchie this is not something I broke. I noticed IDE shows same symptom I fixed before - https://github.com/dotnet/roslyn/pull/15812 - so dug in and saw one of IDE analyzers not implementing IBuiltInAnalyzer which used to implement it (CSharpInlineDeclarationDiagnosticAnalyzer.cs). so I go ahead and search all IDE analyzers and found some of them not implementing IBuiltInAnalyzer so, I fixed them all.

not sure in RC2, what shape it was in. I do know this fix - https://github.com/dotnet/roslyn/pull/15812 - was for RC3.

Pilchie commented:
@CyrusNajmabadi can you investigate the state of RC.2 wrt `IBuiltinAnalyzer` and build and do some buddy testing of our analyzers locally so that we understand this change better?

CyrusNajmabadi commented:
I think there's a little confusion here.  I don't think anything 'broke' per se.  i.e. it's not like things were working, and then they stopped working.  rather, things never worked properly wrt to a few of htese analyzers.  

Basically, an IDE analyzer that just has hidden diagnostics *must* implement IBuiltInAnalyzer.  Or an analyzer that changes severity *must* implmeent IBuiltInAnalyzer.  This was an unknown requirement that was causing problems as we added new analyzers and people used them more.

Heejae is addressing things by trying to ensure that many errant analyzers are not exporting the right information that the diagnostic engine needs.  

--

Heejae, is that correct?

heejaechang commented:
@Pilchie @CyrusNajmabadi only thing this change does as adding IBuiltInAnalyzer to all IDE analyzers. just to be clear.

CyrusNajmabadi commented:
> @CyrusNajmabadi can you investigate the state of RC.2 wrt IBuiltinAnalyzer and build and do some buddy testing of our analyzers locally so that we understand this change better?

Yes.

@heejaechang What should be the *user-observable* change of behavior caused by this PR?  How would i verify that things were now working properly?

Pilchie commented:
@heejaechang but what is the *customer visible effect* of that change.  That's what shiproom is going to care about.  If we can't articulate it, there is no way this change will be accepted.

CyrusNajmabadi commented:
> only thing this change does as adding IBuiltInAnalyzer to all IDE analyzers. just to be clear.

Have you actually verified that Heejae?  Specifically that this is added to "all" IDE analyzers.  Or did you jsut look at the all the ones taht subclass AbstractCodeStyleDiagnosticAnalyzer?

heejaechang commented:
I am not sure how this " This was an unknown requirement that was causing problems as we added new analyzers and people used them more." was unknown requirement? I changed the engine based on the discussion we had on IBuiltInAnalyzer @Pilchie @dpoeschl 

without this PR, try open any file that has CSharpInlineDeclarationDiagnosticAnalyzer, and close it. diagnostic will left out there which should be go away.

CyrusNajmabadi commented:
> I am not sure how this " This was an unknown requirement that was causing problems as we added new analyzers and people used them more." was unknown requirement?

It was unknown because the team didn't know about it :)   That's why so many features got added that didn't use it, and neither the developer nor the reviewers realized it was an issue.

> without this PR, try open any file that has CSharpInlineDeclarationDiagnosticAnalyzer, and close it. diagnostic will left out there which should be go away.

Ok cool.   So i should check all the analyzers you touched to make sure this works properly.

heejaechang commented:
@CyrusNajmabadi I searched one that derived from AbstractCodeStyleDiagnosticAnalyzer and also searched one that implements IBuiltInAnalyzer. but ya, there could be still ones I missed though.

CyrusNajmabadi commented:
Ok.  @Pilchie We'll need to audit it all then.

CyrusNajmabadi commented:
I can do this later tonight, or tomorrow.

CyrusNajmabadi commented:
Ok.  I've validated that all our analyzers now implement IBuiltInAnalyzer

CyrusNajmabadi commented:
Ok.  Have verified that the errors go away after the document closes.  However, it does take about 5 seconds.  Should we not rerun things more quickly on file close so that the errors don't stay around so long.  Maybe 2 seconds would be better?

heejaechang commented:
retest windows_release_unit64_prtest please

heejaechang commented:
yes. right now it is 5 second until task that clean it up runs. making it shorter than that will require some work. so if we want shorter than that, please open an issue.

heejaechang commented:
@MattGertz can I get approval for RTM? 

tagging @Pilchie

MattGertz commented:
I'm with @pilchie on this -- I don't understand the customer visible effect here.  The template needs much more description.

CyrusNajmabadi commented:
@MattGertz I have updated the template.  The crux of the issue is that the current situation feels very noisy and intrusive to the user.  THe more files they edit, the more 'code fix' messages they get, even if they end up closing their files to go work on something else.

The fix makes it so that code-fix messages only apply to the files that are actually open for editing.  The messages will go away when the user closes files to go work on other things.  

We've gotten enough feedback internally and externally to know that this is important.  We've had to scale back on some of the analyzers we've enabled because it ends up causing too high a volume of messages to the user.  THe current situation just exacerbates this as we're not even cleaning up the messages properly when the user goes to work on other things.

MattGertz commented:
OK, because of Dogfooding feedback.  RTM.
